File: modules/M2-capacita/solution/python/app/store.py
from dataclasses import dataclass, replace
from typing import Literal

Status = Literal["todo", "done"]


@dataclass(frozen=True)
class TaskItem:
    id: int
    title: str
    status: Status


class TaskStore:
    def __init__(self) -> None:
        self._tasks: list[TaskItem] = []
        self._next_id = 1

    def all(self) -> list[TaskItem]:
        return list(self._tasks)

    def create(self, title: str) -> TaskItem:
        task = TaskItem(id=self._next_id, title=title, status="todo")
        self._next_id += 1
        self._tasks.append(task)
        return task

    def update_status(self, task_id: int, status: Status) -> TaskItem | None:
        for i, t in enumerate(self._tasks):
            if t.id == task_id:
                updated = replace(t, status=status)
                self._tasks[i] = updated
                return updated
        return None

File: modules/M1-istruzioni/solution/typescript/src/tasks/store.ts
export type TaskStatus = "todo" | "done";

export interface TaskItem {
  id: number;
  title: string;
  status: TaskStatus;
}

export class TaskStore {
  private tasks: TaskItem[] = [];
  private nextId = 1;

  all(): TaskItem[] {
    return [...this.tasks];
  }

  create(title: string): TaskItem {
    const task: TaskItem = { id: this.nextId++, title, status: "todo" };
    this.tasks.push(task);
    return task;
  }

  updateStatus(id: number, status: TaskStatus): TaskItem | null {
    const idx = this.tasks.findIndex((t) => t.id === id);
    if (idx < 0) return null;
    this.tasks[idx] = { ...this.tasks[idx], status };
    return this.tasks[idx];
  }
}
